src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java

```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
+                return "Headers different";
+            }
+            if (!comparison.areEqual()) {
+                return "SamComparison failed";
+            }
+        }
+
+        //At this point we know that the files are not byte-wise identical, but are equal according to SamComparison and their headers are equal
+        //So we iterate over reads and compare them one by one.
+        try(final SamReader reader1 = getReader(sam1, validation, reference);
+            final SamReader reader2 = getReader(sam2, validation, reference)) {
+            final SAMRecordIterator it1 = reader1.iterator();
+            final SAMRecordIterator it2 = reader2.iterator();
+            while (it1.hasNext() && it2.hasNext()) {
+                final SAMRecord read1 = it1.next();
+                final SAMRecord read2 = it2.next();
+                final String eqMessage = readsEqualAllowAddingAttributes(read1, read2);
+                if (eqMessage != null){
+                    return eqMessage;
+                }
+            }
+            if (it1.hasNext() || it2.hasNext()) {
+                //at least one has no more records (because the while loop is done) and at least one does have records. So we're not equal.
+                return "Not the same number of reads";
+            }
+            return null;
+        }
+    }
+
+    /**
+     * Compares the reads but ignores order or attributes.
+     * Also allows read2 to have a superset of attributes of read1.
+     */
+    private static String readsEqualAllowAddingAttributes(final SAMRecord read1, final SAMRecord read2) {
+        final SAMRecordCoordinateComparator coordinateComparator = new SAMRecordCoordinateComparator();
+        if (!Objects.equals(read1.getReadName(), read2.getReadName())){
+            return "name different read1:" + read1.getReadName() + " read2:" + read2.getReadName() ;
+        }
+        if (!Objects.equals(read1.getFlags(), read2.getFlags())){
+            return "flags different read1:" + read1.getReadName() + " read2:" + read2.getReadName() ;
+        }
+        if (!Objects.equals(read1.getInferredInsertSize(), read2.getInferredInsertSize())){
+            return "getInferredInsertSize different read1:" + read1.getReadName() + " read2:" + read2.getReadName() + " (" + read1.getInferredInsertSize()+ " vs " + read2.getInferredInsertSize() + ")";
+        }
+        if (!Objects.equals(read1.getMappingQuality(), read2.getMappingQuality())){
+            return "getMappingQuality different read1:" + read1.getReadName() + " read2:" + read2.getReadName() ;
+        }
+        //Note the comparator does check flags, MQ and insert size but we want better messages
+        if (0 != coordinateComparator.fileOrderCompare(read1, read2)){
+            return "file order read1:" + read1.getContig() + ":" + read1.getAlignmentStart() + " read2:" + read2.getContig() + ":" + read2.getAlignmentStart();
```

Comment:
Don't use a `SAMRecordCoordinateComparator` here -- it potentially duplicates a lot of comparisons. Instead, just check directly that the reference index/name and alignment start are equal, and adds checks here for the mate position as well.